File: src/main/java/dev/tuhkanens/comfortlib/database/base/DatabaseBase.java
package dev.tuhkanens.comfortlib.database.base;

import org.jdbi.v3.core.Jdbi;

public abstract class DatabaseBase {

    private Jdbi database;

    protected abstract Jdbi createConnection();
    protected void closeConnection() {}

    public void connect() {
        database = createConnection();
    }

    public void disconnect() {
        if (database == null) return;
        closeConnection();
    }

    public Jdbi getDatabase() {
        return database;
    }

}

File: src/main/kotlin/dev/tuhkanens/comfortlib/database/DatabaseBase.kt
package dev.tuhkanens.comfortlib.database

import org.jetbrains.exposed.v1.jdbc.Database

abstract class DatabaseBase {

    private lateinit var database: Database

    protected abstract fun createConnection(): Database
    protected open fun closeConnection() {}

    fun connect() {
        database = createConnection()
    }

    fun disconnect() {
        if (!::database.isInitialized) return
        closeConnection()
    }

    fun getDatabase(): Database {
        return database
    }

}